lib/rubocop/cop/rspec/nested_groups.rb

```diff
@@ -97,27 +97,25 @@ class NestedGroups < Cop
           "Configuration key `#{DEPRECATED_MAX_KEY}` for #{cop_name} is " \
           'deprecated in favor of `Max`. Please use that instead.'
 
-        def_node_search :find_contexts, ExampleGroups::ALL.block_pattern
-
         def on_top_level_describe(node, _args)
-          find_nested_contexts(node.parent) do |context, nesting|
+          find_nested_example_groups(node.parent) do |example_group, nesting|
             self.max = nesting
             add_offense(
-              context.send_node,
+              example_group.send_node,
               message: message(nesting)
             )
           end
         end
 
         private
 
-        def find_nested_contexts(node, nesting: 1, &block)
-          find_contexts(node) do |nested_context|
-            yield(nested_context, nesting) if nesting > max_nesting
+        def find_nested_example_groups(node, nesting: 1, &block)
+          yield node, nesting if example_group?(node) && nesting > max_nesting
+
+          next_nesting = example_group?(node) ? nesting + 1 : nesting
 
-            nested_context.each_child_node do |child|
-              find_nested_contexts(child, nesting: nesting + 1, &block)
-            end
+          node.each_child_node do |child|
```

Comment:
It might be possible to squeeze more out of this with `node.each_descendant(:block)`.
Which one would perform better?

Reply:
Filtering by node type looks reasonable (but `each_descendant` isn't suitable here because it's recursive but we need here direct children).

There is again one issue. We should handle not only `block` but at least `begin` node as well - if there are several nested `context`s they are wrapped into `begin` node. TBH I am not sure whether there are another edge cases (e.g. `kwbegin`) or we could just filter `block` and `begin` nodes:

```ruby
node.each_child_node(:block, :begin) do |child|
```

Reply:
It has impressive impact - decreases the cop share from 3.3% to 1.8%.

There is no any difference between generated offenses with and without with filtering (checked on GitLab specs) but I still hesitate.

Reply:
I'm a bit lost. What change reduced the share from 3.3% to 1.8%?

Reply:
I meant additional filtering of `block` and `begin` nodes (what you've proposed) decreased cop time from 3.3% to 1.8%.

Reply:
ping @Darhazer @bquorning What do you think?

Reply:
Yeah, I think using `node.each_child_node(:block, :begin)` is fine. If we are missing some edge cases, we can fix them as bugs. Provided of course that anyone will discover them.

Reply:
Done